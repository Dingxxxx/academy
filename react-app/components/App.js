var React = require('react');
var xhr = require('xhr');
var connect = require('react-redux').connect;

var Plot = require('./Plot');
var actions = require('../actions');

var App = React.createClass({
  fetchData: function(evt) {
    evt.preventDefault();

    var location = encodeURIComponent(this.props.location);

    var urlPrefix = 'http://api.openweathermap.org/data/2.5/forecast?q=';
    var urlSuffix = '&APPID=dbe69e56e7ee5f981d76c3e77bbb45c0&units=metric';
    var url = urlPrefix + location + urlSuffix;

    this.props.dispatch(actions.loadData(url));
  },
  onPlotClick: function(data) {
    if (data.points) {
      var number = data.points[0].pointNumber;
      this.props.dispatch(actions.setSelectedDate(this.props.dates[number]));
      this.props.dispatch(actions.setSelectedTemp(this.props.temps[number]))
    }
  },
  changeLocation: function(evt) {
    this.props.dispatch(actions.setLocation(evt.target.value));
  },
  render: function() {
    var currentTemp = 'not loaded yet';
    if (this.props.data.list) {
      currentTemp = this.props.data.list[0].main.temp;
    }
    return (
      <div>
        <h1>Weather</h1>
        <form onSubmit={this.fetchData}>
          <label>I want to know the weather for
            <input
              placeholder={"City, Country"}
              type="text"
              value={this.props.location}
              onChange={this.changeLocation}
            />
          </label>
        </form>
        {/*
          Render the current temperature and the forecast if we have data
          otherwise return null
        */}
        {(this.props.data.list) ? (
          <div className="wrapper">
            {/* Render the current temperature if no specific date is selected */}
            <p className="temp-wrapper">
              <span className="temp">
                { this.props.selected.temp ? this.props.selected.temp : currentTemp }
              </span>
              <span className="temp-symbol">Â°C</span>
              <span className="temp-date">
                { this.props.selected.temp ? this.props.selected.date : ''}
              </span>
            </p>
            <h2>Forecast</h2>
            <Plot
              xData={this.props.dates}
              yData={this.props.temps}
              onPlotClick={this.onPlotClick}
              type="scatter"
            />
          </div>
        ) : null}

      </div>
    );
  }
});

module.exports = connect(function (state) {
  return state;
})(App);
