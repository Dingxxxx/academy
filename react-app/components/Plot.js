// components/Plot.js
var React = require('react');

var Plot = React.createClass({
  drawPlot: function() {
    Plotly.newPlot('plot', [{
      x: this.props.xData,
      y: this.props.yData,
      type: this.props.type
    }]);
    this.refs.plot.on('plotly_click', this.props.onPlotClick);
  },
  componentDidMount: function() {
    this.drawPlot();
  },
  componentDidUpdate: function() {
    this.drawPlot();
  },
  render: function() {
    return (
      <div id="plot" ref="plot"></div>
    );
  }
});

module.exports = Plot;
