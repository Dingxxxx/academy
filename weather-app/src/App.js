import React, { Component } from 'react';
import xhr from 'xhr';

class App extends Component {
  state = {
    location: '',
    data: {}
  };

  fetchData = (evt) => {
    evt.preventDefault();

    var location = encodeURIComponent(this.state.location);

    var urlPrefix = 'http://api.openweathermap.org/data/2.5/forecast?q=';
    var urlSuffix = '&APPID=dbe69e56e7ee5f981d76c3e77bbb45c0&units=metric';
    var url = urlPrefix + location + urlSuffix;

    var self = this;

    xhr({
      url: url
    }, function (err, data) {
      var data = JSON.parse(data.body);
      self.setState({
        data: data
      });
    });
  };

  changeLocation = (evt) => {
    this.setState({
      location: evt.target.value
    });
  };

  render() {
    console.log(this.state);
    var currentTemp = 'not loaded yet';
    if (this.state.data.list) {
      currentTemp = this.state.data.list[0].main.temp;
    }
    return (
      <div>
        <h1>Weather</h1>
        <form onSubmit={this.fetchData}>
          <label>I want to know the weather for
            <input
              placeholder={"City, Country"}
              type="text"
              value={this.state.location}
              onChange={this.changeLocation}
            />
          </label>
        </form>
        <p className="temp-wrapper">
          <span className="temp">{ currentTemp }</span>
          <span className="temp-symbol">Â°C</span>
        </p>
      </div>
    );
  }
}

export default App;
