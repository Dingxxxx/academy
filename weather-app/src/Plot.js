// Plot.js
import React, { Component } from 'react';

class Plot extends Component {
  drawPlot = () => {
    Plotly.newPlot('plot', [{
      x: this.props.xData,
      y: this.props.yData,
      type: this.props.type
    }], {
      margin: {
        t: 0, r: 0, l: 30
      },
      xaxis: {
        gridcolor: 'transparent'
      }
    }, {
      displayModeBar: false
    });
    this.refs.plot.on('plotly_click', this.props.onPlotClick);
  }
  componentDidUpdate() {
    this.drawPlot();
  }

  componentDidMount() {
    this.drawPlot();
  }

  render() {
    return (
      <div id="plot" ref="plot"></div>
    );
  }
}

export default Plot;
