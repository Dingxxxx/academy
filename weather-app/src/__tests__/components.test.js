import React from 'react';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';
import { AppÂ } from '../App';
import Plot from '../Plot';

describe('components', () => {
	describe('<App />', () => {
		it('renders correctly', () => {
			const tree = renderer.create(<App redux={fromJS({})} />).toJSON();
		  expect(tree).toMatchSnapshot();
		});
	});

	describe('<Plot />', () => {
		it('renders correctly', () => {
			const tree = renderer.create(<Plot />).toJSON();
		  expect(tree).toMatchSnapshot();
		});
	});
});
