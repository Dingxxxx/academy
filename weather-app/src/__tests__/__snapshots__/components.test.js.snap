exports[`components <App /> renders correctly 1`] = `
<div>
  <h1>
    Weather
  </h1>
  <form
    onSubmit={[Function anonymous]}>
    <label>
      I want to know the weather for
      <input
        onChange={[Function anonymous]}
        placeholder="City, Country"
        type="text"
        value={undefined} />
    </label>
  </form>
</div>
`;
